pkg/engine/internal/workflow/runner.go

```diff
@@ -0,0 +1,144 @@
+package workflow
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/grafana/loki/v3/pkg/engine/internal/executor"
+)
+
+// A Runner can asynchronously execute a workflow.
+type Runner interface {
+	// AddStreams registers a list of Streams that can be used by Tasks.
+	// AddStreams returns an error if any of the streams (by ID) are already
+	// registered.
+	//
+	// The provided handler will be called whenever any of the provided streams
+	// change state.
+	AddStreams(ctx context.Context, handler StreamEventHandler, streams ...*Stream) error
```

Comment:
Wdyt about a different naming for streams (some options: edges, channels, links, pipes, connections)? I'm wondering if `Stream` as a "connection between the tasks" and `Stream` as "log stream" might become confusing 🤔